### DISCUSSION TITLE
How to manage schema in teamwork?

### DISCUSSION THREAD

**Discussion by pizza7311** [prisma/prisma#23997] [Q&A]
### Question

Hello. i am using mysql with prisma while developing with my team.
Our team uses one database to develop each feature. and only uses `prisma db push` to synchronize schema.
The problem is when each teammates are pushing schema to database previous works is overwritten

For example here's the original schema.
```prisma
model A {
 //...
}

model B {
 //...
}
```

and teammates make each other's branches and update schema

this is my schema
```prisma
//...

model C {
 //add new model
}
```

and this is team's schema
```prisma
//...

model D {
 //...
}
```

If i push the schema, prisma will create a new table. after that, my teammate finish his work and push schema to database and my works will overwritten. i haven't used `prisma migration dev`, but can migration solve this problem? or there's any other solution?

**Comment by ludralph** [comment]
Hi @pizza7311 👋 

Using Prisma Migrate can help solve this problem. Prisma Migrate is a database schema migration tool that enables you to keep your database schema in sync with your Prisma schema, and version control for your database schema.

When you make changes to your Prisma schema, you can generate a new migration with Prisma Migrate. This migration can be committed to your version control system (like Git), allowing your teammates to pull the changes and apply the migration to their local databases. This way, each teammate's changes can be applied sequentially without overwriting each other's work.

Here's a basic workflow for incorporating changes from collaborators:

1. Pull the changed Prisma schema and ./prisma/migrations folder
2. Run the migrate dev command to apply new migrations:
```
npx prisma migrate dev
```
Migrations are applied in the same order as they were created. The creation date is part of the migration subfolder name - for example, `20210316081837-updated-fields` was created on `2021-03-16-08:08:37`.

Please note that this guide does not apply for MongoDB. Instead of `migrate dev`, db push is used for MongoDB. For more details, you can refer to the [Prisma documentation on team development with Prisma Migrate](https://www.prisma.io/docs/orm/prisma-migrate/workflows/team-development).

**Reply by lroal** [reply]
But this still means each team must be synchronized with a shared migration branch ? Right ?
The cannot create independent migrations in parallel branches , or ?
Because migrations must sequential and not parallel.

**Reply by ludralph** [reply]
Hi @lroal 

You're correct in your understanding. Using Prisma Migrate does require some level of synchronization between team members, especially when it comes to migrations. 

- Sequential Migrations: As you pointed out, migrations are indeed sequential. Each migration builds upon the previous state of the database. This is by design to ensure data integrity and consistency.

- Shared Migration History: The team needs to maintain a shared migration history. This typically means having a main branch (like main or develop) where all finalized migrations are merged.

- Parallel Development Challenges: Creating completely independent migrations in parallel branches can indeed be challenging. If two team members create migrations simultaneously on separate branches, merging these can lead to conflicts.
[Prisma Docs: Team development with Prisma Migrate](https://www.prisma.io/docs/orm/prisma-migrate/workflows/team-development) addresses this issue:

> When working on separate features in parallel, team members might create migrations that conflict with each other when merged. This is because Prisma Migrate expects migrations to be applied sequentially."

However, there are strategies to mitigate these challenges:

- Short-lived Feature Branches: Encourage team members to work on short-lived feature branches and merge frequently to minimize conflicts.

- Communication: Team members should communicate about schema changes to avoid conflicting modifications.

- Resolving Conflicts: When conflicts occur, they need to be resolved manually. This might involve squashing migrations or carefully merging changes.

- Using prisma migrate dev: This command can help detect and resolve conflicts when pulling changes from the main branch.
It's worth noting that while this workflow requires more coordination, it provides better control and visibility over database schema changes compared to using db push.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
